paz/models/detection/frustum_pointnet.py

```diff
@@ -0,0 +1,177 @@
+
+    logits = Instance_Seg_Net(point_cloud, one_hot_vec)  # bs,n,2
+    end_points['mask_logits'] = logits
+
+    # Mask Point Centroid
+    object_point_cloud_xyz, mask_xyz_mean, end_points = point_cloud_masking(point_cloud, logits, end_points)
+
+    # T-Net
+    center_delta = TNet(object_point_cloud_xyz, one_hot_vec)  # (32,3)
+
+    stage1_center = center_delta + mask_xyz_mean  # Bx3
+    end_points['stage1_center'] = stage1_center
+    # Get object point cloud in object coordinate
+    object_point_cloud_xyz_new = object_point_cloud_xyz - tf.expand_dims(center_delta, 1)
+
+    # 3D Box Estimation
+    box_pred = Box_Est_Net(object_point_cloud_xyz_new, one_hot_vec)  # (32, 59)
+
+    end_points = parse_output_to_tensors(box_pred, end_points)
+    end_points['center'] = end_points['center_boxnet'] + stage1_center  # Bx3
+
+    logits = end_points['mask_logits']
+    mask = end_points['mask']
+    stage1_center = end_points['stage1_center']
+    center_boxnet = end_points['center_boxnet']
+    heading_scores = end_points['heading_scores']  # BxNUM_HEADING_BIN
+    heading_residuals_normalized = end_points['heading_residuals_normalized']
+    heading_residuals = end_points['heading_residuals']
+    size_scores = end_points['size_scores']
+    size_residuals_normalized = end_points['size_residuals_normalized']
+    size_residuals = end_points['size_residuals']
+    center = end_points['center']
+
+    logits = Lambda(lambda x: x, name="InsSeg_out")(logits)
+    box3d_center = Lambda(lambda x: x, name="center_out")(center)
+    heading_scores = Lambda(lambda x: x, name="heading_scores")(heading_scores)
+    heading_residual = Lambda(lambda x: x, name="heading_residual")(heading_residuals)
+    heading_residuals_normalized = Lambda(lambda x: x, name="heading_residual_norm")(heading_residuals_normalized)
+    size_scores = Lambda(lambda x: x, name="size_scores")(size_scores)
+    size_residual = Lambda(lambda x: x, name="size_residual")(size_residuals)
+    size_residuals_normalized = Lambda(lambda x: x, name="size_residual_norm")(size_residuals_normalized)
+
+    loss = Lambda(FPointNet_loss, output_shape=(1,), name='fp_loss',
+                  arguments={'corner_loss_weight': 10.0, 'box_loss_weight': 1.0})([mask_label, center_label,
+                                                                                   heading_class_label,
+                                                                                   heading_residual_label,
+                                                                                   size_class_label,
+                                                                                   size_residual_label, end_points])
+
+    training_model = Model([point_cloud, one_hot_vec, mask_label, center_label, heading_class_label,
+                            heading_residual_label, size_class_label, size_residual_label], loss,
+                           name='f_pointnet_train')
```

Comment:
maybe it would be cleaner to name a variable inputs=[point_cloud, one_hot_vector, ....] and then pass it to the model creation rather than passing all these variable names in one go. It makes it difficult to read what's input and what's output.